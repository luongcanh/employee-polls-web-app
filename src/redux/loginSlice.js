import { createSlice } from '@reduxjs/toolkit';

const initialState = {};

export const loginSlice = createSlice({
	name: 'login',
	initialState,
	reducers: {
		setLogin: (state, action) => {
			state.authedUser = action.payload.id;
		},
		setLogout: (state) => {
			state.authedUser = null;
		}
	},
});

export const { setLogin, setLogout } = loginSlice.actions;

export const selectLogin = (state) => state?.login?.authedUser;

export default loginSlice.reducer;
