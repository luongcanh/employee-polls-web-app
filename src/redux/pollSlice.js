import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { getAllPolls } from './utils/callAPI';
import { ANSWER_OPTION } from '../constants/constant';
import { checkIfAnswered } from './utils/common';
import { _saveQuestion, _saveQuestionAnswer } from '../_DATA';

const initialState = {
	loading: false,
	polls: [],
};

/**
 * @description Get all polls
 */
export const fetchPolls = createAsyncThunk('polls/getAllPolls', async () => await getAllPolls());

export const handleSubmitAnswer = createAsyncThunk(
	'polls/handleSubmitAnswer',
	async ({ authedUser, qid, answer }) => {
		await _saveQuestionAnswer({ authedUser, qid, answer });

		return { authedUser, qid, answer };
	},
);

/**
 * @description Create new poll
 */
export const handleCreateQuestion = createAsyncThunk(
	'polls/handleCreateQuestion',
	async ({ author, optionOneText, optionTwoText }) => {
		const result = await _saveQuestion({ author, optionOneText, optionTwoText });

		return result;
	},
);

const pollSlice = createSlice({
	name: 'polls',
	initialState,
	reducers: {},
	extraReducers: (builder) => {
		builder.addCase(fetchPolls.pending, (state) => {
			state.loading = true;
		});
		builder.addCase(fetchPolls.fulfilled, (state, action) => {
			state.polls = action.payload;
			state.loading = false;
		});
		builder.addCase(handleSubmitAnswer.pending, (state) => {
			state.loading = true;
		});
		builder.addCase(handleSubmitAnswer.fulfilled, (state, action) => {
			let currentVoters = state.polls[action.payload.qid][action.payload.answer].votes;
			if (
				!checkIfAnswered(state.polls[action.payload.qid], action.payload.authedUser) &&
				(action.payload.answer === ANSWER_OPTION.OPTION_ONE ||
					action.payload.answer === ANSWER_OPTION.OPTION_TWO)
			) {
				currentVoters.push(action.payload.authedUser);
			}
			state.loading = false;
		});
		builder.addCase(handleCreateQuestion.pending, (state) => {
			state.loading = true;
		});
		builder.addCase(handleCreateQuestion.fulfilled, (state, action) => {
			state.loading = false;
			state.polls[action.payload.id] = action.payload;
		});
		builder.addCase(handleCreateQuestion.rejected, (state, action) => {
			state.loading = false;
		});
	},
});

export default pollSlice.reducer;
