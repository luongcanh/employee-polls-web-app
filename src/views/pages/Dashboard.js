import React, { useEffect, useState } from 'react';
import { Box, Tab, Tabs, Container } from '@mui/material';
import QuestionsList from '../components/QuestionsList';
import { useSelector } from 'react-redux';
import { ANSWER_TYPE } from '../../constants/constant';

const Dashboard = () => {
	const [value, setValue] = useState(0);
	const authedUser = useSelector((state) => state.login.authedUser);
	const [newQuestions, setNewQuestions] = useState({});
	const [answeredQuestions, setAnsweredQuestions] = useState({});
	const allQuestions = useSelector((state) => state.polls.polls);
	const users = useSelector((state) => state.users.users);

	useEffect(() => {
		if (authedUser) {

			const answeredQuestionIds = Object.keys(users[authedUser].answers).sort(
				(firstQId, secondQId) => {
					return allQuestions[secondQId].timestamp - allQuestions[firstQId].timestamp;
				},
			);

			const newQuestionIds = Object.keys(allQuestions)
				.filter((questionId) => !users[authedUser].answers[questionId])
				.sort((firstQId, secondQId) => {
					return allQuestions[secondQId].timestamp - allQuestions[firstQId].timestamp;
				});
			
			setNewQuestions(
				newQuestionIds.reduce((acc, questionId) => {
					return { ...acc, [questionId]: allQuestions[questionId] };
				}, {}),
			);
			
			setAnsweredQuestions(
				answeredQuestionIds.reduce((acc, questionId) => {
					return { ...acc, [questionId]: allQuestions[questionId] };
				}, {}),
			);
		}
		// eslint-disable-next-line
	}, []);

	/**
	 * @description Handle tab selection changes (New Question/ Done)
	 */
	const handleOnChange = (event, newValue) => {
		setValue(newValue);
	};

	return (
		<div>
			<Container maxWidth="xl">
				<Box sx={{ width: '100%' }}>
					<Box sx={{ borderBottom: 1, borderColor: 'divider' }}>
						<Tabs value={value} onChange={handleOnChange}>
							<Tab label="New Question" />
							<Tab label="Done" />
						</Tabs>
					</Box>
					{value === 0 && (
                        <QuestionsList questions={newQuestions} questionType={ANSWER_TYPE.UNANSWERED} />
                    )}
                    {value === 1 && (
                        <QuestionsList questions={answeredQuestions} questionType={ANSWER_TYPE.ANSWERED} />
                    )}
				</Box>
			</Container>
		</div>
	);
};

export default Dashboard;
