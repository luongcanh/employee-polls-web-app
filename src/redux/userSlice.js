import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { getAllUsers } from './utils/callAPI';
import { handleCreateQuestion, handleSubmitAnswer } from './pollSlice';

const initialState = {
    loading: false,
    users: [],
};

export const fetchUsers = createAsyncThunk('users/getAllUsers', async () => await getAllUsers());

const userSlice = createSlice({
    name: 'users',
    initialState,
    reducers: {},
    extraReducers: (builder) => {
        builder.addCase(fetchUsers.pending, (state) => {
            state.loading = true;
        });
        builder.addCase(fetchUsers.fulfilled, (state, action) => {
            state.users = action.payload;
            state.loading = false;
        });
        builder.addCase(handleSubmitAnswer.pending, (state) => {
            state.loading = true;
        });
        builder.addCase(handleSubmitAnswer.fulfilled, (state, action) => {
            state.users[action.payload.authedUser].answers[action.payload.qid] = action.payload.answer;
            state.loading = false;
        });
        builder.addCase(handleCreateQuestion.pending, (state) => {
            state.loading = true;
        });
        builder.addCase(handleCreateQuestion.fulfilled, (state, action) => {
            state.loading = false;
            state.users[action.payload.author].questions.push(action.payload.id);
        });
        builder.addCase(handleCreateQuestion.rejected, (state, action) => {
            state.loading = false;
        });
    },
});

export default userSlice.reducer;
