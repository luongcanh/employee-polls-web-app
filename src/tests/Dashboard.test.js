/* eslint-disable testing-library/prefer-screen-queries */
import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import Dashboard from '../views/pages/Dashboard';
import { store } from '../redux/store';
import { setLogin } from '../redux/loginSlice';
import { fetchPolls } from '../redux/pollSlice';
import { fetchUsers } from '../redux/userSlice';
import { BrowserRouter as Router } from 'react-router-dom';

describe('Dashboard Component', () => {
	let component;

	beforeEach(async () => {
		await store.dispatch(fetchPolls());
		await store.dispatch(fetchUsers());
		await store.dispatch(setLogin({ id: 'sarahedo' }));
		// eslint-disable-next-line
		component = render(
			<Provider store={store}>
				<Router>
					<Dashboard />
				</Router>
			</Provider>,
		);
	});

	it('renders all tabs correctly', () => {
        expect(component).toMatchSnapshot();
	});

	it('renders New Question tab content correctly', () => {
		const { getByText, queryAllByText } = component;
		fireEvent.click(getByText('New Question'));
		const noNewQuestions =
			Object.keys(store.getState().polls.polls).length -
			Object.keys(store.getState().users.users['sarahedo'].answers).length;
		expect(queryAllByText('Show').length).toEqual(noNewQuestions);
	});

	it('renders Done tab content correctly', () => {
		const { getByText, queryAllByText } = component;
		fireEvent.click(getByText('Done'));
		const noAnswerQuestions = Object.keys(store.getState().users.users['sarahedo'].answers).length;
		expect(queryAllByText('Show').length).toEqual(noAnswerQuestions);
	});
});
