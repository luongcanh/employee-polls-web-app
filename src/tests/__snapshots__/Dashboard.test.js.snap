// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Component renders all tabs correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="MuiContainer-root MuiContainer-maxWidthXl css-19r6kue-MuiContainer-root"
        >
          <div
            class="MuiBox-root css-8atqhb"
          >
            <div
              class="MuiBox-root css-1gsv261"
            >
              <div
                class="MuiTabs-root css-1ujnqem-MuiTabs-root"
              >
                <div
                  class="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
                  style="overflow: hidden; margin-bottom: 0px;"
                >
                  <div
                    class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
                    role="tablist"
                  >
                    <button
                      aria-selected="true"
                      class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
                      role="tab"
                      tabindex="0"
                      type="button"
                    >
                      New Question
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </button>
                    <button
                      aria-selected="false"
                      class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
                      role="tab"
                      tabindex="-1"
                      type="button"
                    >
                      Done
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </button>
                  </div>
                  <span
                    class="MuiTabs-indicator css-1aquho2-MuiTabs-indicator"
                    style="left: 0px; width: 0px;"
                  />
                </div>
              </div>
            </div>
            <div>
              <div
                class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
              >
                <h3
                  class="MuiTypography-root MuiTypography-h3 css-1uku6ty-MuiTypography-root"
                >
                  New Questions
                </h3>
                <div
                  class="MuiBox-root css-qgh0pp"
                >
                  <div
                    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation5 css-qh7kl5-MuiPaper-root"
                  >
                    <div
                      class="MuiBox-root css-12dv9jb"
                    >
                      <div
                        class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
                      >
                        <div
                          class="MuiCardContent-root css-o8l425-MuiCardContent-root"
                        >
                          <h5
                            class="MuiTypography-root MuiTypography-h5 css-1jcm1rz-MuiTypography-root"
                          >
                            mtsamis
                          </h5>
                          <p
                            class="MuiTypography-root MuiTypography-body1 css-9eg9l8-MuiTypography-root"
                          >
                            02:16 | 01-05-2017
                          </p>
                        </div>
                        <div
                          class="MuiCardActions-root MuiCardActions-spacing css-t2o2tm-MuiCardActions-root"
                        >
                          <button
                            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSuccess MuiButton-sizeLarge MuiButton-outlinedSizeLarge MuiButton-colorSuccess MuiButton-root MuiButton-outlined MuiButton-outlinedSuccess MuiButton-sizeLarge MuiButton-outlinedSizeLarge MuiButton-colorSuccess css-12wndg-MuiButtonBase-root-MuiButton-root"
                            tabindex="0"
                            type="button"
                          >
                            Show
                            <span
                              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                            />
                          </button>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation5 css-qh7kl5-MuiPaper-root"
                  >
                    <div
                      class="MuiBox-root css-12dv9jb"
                    >
                      <div
                        class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
                      >
                        <div
                          class="MuiCardContent-root css-o8l425-MuiCardContent-root"
                        >
                          <h5
                            class="MuiTypography-root MuiTypography-h5 css-1jcm1rz-MuiTypography-root"
                          >
                            tylermcginnis
                          </h5>
                          <p
                            class="MuiTypography-root MuiTypography-body1 css-9eg9l8-MuiTypography-root"
                          >
                            19:09 | 15-03-2017
                          </p>
                        </div>
                        <div
                          class="MuiCardActions-root MuiCardActions-spacing css-t2o2tm-MuiCardActions-root"
                        >
                          <button
                            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSuccess MuiButton-sizeLarge MuiButton-outlinedSizeLarge MuiButton-colorSuccess MuiButton-root MuiButton-outlined MuiButton-outlinedSuccess MuiButton-sizeLarge MuiButton-outlinedSizeLarge MuiButton-colorSuccess css-12wndg-MuiButtonBase-root-MuiButton-root"
                            tabindex="0"
                            type="button"
                          >
                            Show
                            <span
                              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                            />
                          </button>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="MuiContainer-root MuiContainer-maxWidthXl css-19r6kue-MuiContainer-root"
      >
        <div
          class="MuiBox-root css-8atqhb"
        >
          <div
            class="MuiBox-root css-1gsv261"
          >
            <div
              class="MuiTabs-root css-1ujnqem-MuiTabs-root"
            >
              <div
                class="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
                style="overflow: hidden; margin-bottom: 0px;"
              >
                <div
                  class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
                  role="tablist"
                >
                  <button
                    aria-selected="true"
                    class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
                    role="tab"
                    tabindex="0"
                    type="button"
                  >
                    New Question
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                  <button
                    aria-selected="false"
                    class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
                    role="tab"
                    tabindex="-1"
                    type="button"
                  >
                    Done
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                </div>
                <span
                  class="MuiTabs-indicator css-1aquho2-MuiTabs-indicator"
                  style="left: 0px; width: 0px;"
                />
              </div>
            </div>
          </div>
          <div>
            <div
              class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
            >
              <h3
                class="MuiTypography-root MuiTypography-h3 css-1uku6ty-MuiTypography-root"
              >
                New Questions
              </h3>
              <div
                class="MuiBox-root css-qgh0pp"
              >
                <div
                  class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation5 css-qh7kl5-MuiPaper-root"
                >
                  <div
                    class="MuiBox-root css-12dv9jb"
                  >
                    <div
                      class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
                    >
                      <div
                        class="MuiCardContent-root css-o8l425-MuiCardContent-root"
                      >
                        <h5
                          class="MuiTypography-root MuiTypography-h5 css-1jcm1rz-MuiTypography-root"
                        >
                          mtsamis
                        </h5>
                        <p
                          class="MuiTypography-root MuiTypography-body1 css-9eg9l8-MuiTypography-root"
                        >
                          02:16 | 01-05-2017
                        </p>
                      </div>
                      <div
                        class="MuiCardActions-root MuiCardActions-spacing css-t2o2tm-MuiCardActions-root"
                      >
                        <button
                          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSuccess MuiButton-sizeLarge MuiButton-outlinedSizeLarge MuiButton-colorSuccess MuiButton-root MuiButton-outlined MuiButton-outlinedSuccess MuiButton-sizeLarge MuiButton-outlinedSizeLarge MuiButton-colorSuccess css-12wndg-MuiButtonBase-root-MuiButton-root"
                          tabindex="0"
                          type="button"
                        >
                          Show
                          <span
                            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                          />
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation5 css-qh7kl5-MuiPaper-root"
                >
                  <div
                    class="MuiBox-root css-12dv9jb"
                  >
                    <div
                      class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
                    >
                      <div
                        class="MuiCardContent-root css-o8l425-MuiCardContent-root"
                      >
                        <h5
                          class="MuiTypography-root MuiTypography-h5 css-1jcm1rz-MuiTypography-root"
                        >
                          tylermcginnis
                        </h5>
                        <p
                          class="MuiTypography-root MuiTypography-body1 css-9eg9l8-MuiTypography-root"
                        >
                          19:09 | 15-03-2017
                        </p>
                      </div>
                      <div
                        class="MuiCardActions-root MuiCardActions-spacing css-t2o2tm-MuiCardActions-root"
                      >
                        <button
                          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSuccess MuiButton-sizeLarge MuiButton-outlinedSizeLarge MuiButton-colorSuccess MuiButton-root MuiButton-outlined MuiButton-outlinedSuccess MuiButton-sizeLarge MuiButton-outlinedSizeLarge MuiButton-colorSuccess css-12wndg-MuiButtonBase-root-MuiButton-root"
                          tabindex="0"
                          type="button"
                        >
                          Show
                          <span
                            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                          />
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
